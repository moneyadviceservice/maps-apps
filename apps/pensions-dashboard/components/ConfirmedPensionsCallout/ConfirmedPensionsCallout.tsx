import { Heading } from '@maps-react/common/components/Heading';
import { InformationCallout } from '@maps-react/common/components/InformationCallout';
import { Link } from '@maps-react/common/components/Link';
import { Paragraph } from '@maps-react/common/components/Paragraph';
import useTranslation from '@maps-react/hooks/useTranslation';

import { PensionType, STATE_RETIREMENT_AGE } from '../../lib/constants';
import { PensionArrangement } from '../../lib/types';
import { getRetirementAge, processBenefitIllustrations } from '../../lib/utils';

type ConfirmedPensionsCalloutProps = {
  data: PensionArrangement;
};

export const ConfirmedPensionsCallout = ({
  data: {
    schemeName,
    employmentMembershipPeriods,
    pensionAdministrator: { name },
    externalAssetId,
    pensionType,
    benefitIllustrations,
    retirementDate,
    dateOfBirth,
  },
}: ConfirmedPensionsCalloutProps) => {
  const { t, locale } = useTranslation();

  const { monthlyAmount } = processBenefitIllustrations(
    benefitIllustrations,
    t,
  );

  return (
    <InformationCallout className="px-6 py-6">
      <Heading
        color="text-blue-800"
        level="h4"
        className="flex gap-2 mt-0 mb-4 items-top"
      >
        {pensionType === PensionType.SP ? schemeName : name}
      </Heading>

      {pensionType !== 'SP' && (
        <>
          <Paragraph className="mb-0 text-gray-400">
            {t('common.scheme')}:
          </Paragraph>
          <Paragraph>{schemeName}</Paragraph>
        </>
      )}

      {pensionType === PensionType.SP && <Paragraph>{name}</Paragraph>}

      {employmentMembershipPeriods?.length && (
        <Paragraph className="mb-0 text-gray-400">
          {t('common.employer')}:
        </Paragraph>
      )}

      {employmentMembershipPeriods?.length &&
        employmentMembershipPeriods.map((employer, idx) => (
          <Paragraph
            className="pb-2 mb-2 border-b-1 border-slate-400"
            key={`${employer.employerName}-${idx}`}
          >
            {employer.employerName}
          </Paragraph>
        ))}

      <Paragraph className="mb-0 text-gray-400">
        {t('common.retirement-age')}
      </Paragraph>

      <Paragraph className="pb-2 mb-2 border-b-1 border-slate-400">
        {/* temporary hardcoding of State Retirement Age because DOB is not available for a SP */}
        {pensionType === PensionType.SP
          ? STATE_RETIREMENT_AGE
          : getRetirementAge(retirementDate, dateOfBirth)}
      </Paragraph>

      <Paragraph className="mb-0 text-gray-400">
        {t('common.estimated-income')}
      </Paragraph>

      <Paragraph className="pb-2 border-b-1 border-slate-400">
        <strong className="mr-2 text-xl">{monthlyAmount}</strong>
        {t('common.a-month')}
      </Paragraph>

      <Link
        asButtonVariant="primary"
        href={`/${locale}/pension-details/${externalAssetId}`}
        data-testid="details-link"
      >
        {t('common.details-link')}
      </Link>
    </InformationCallout>
  );
};
