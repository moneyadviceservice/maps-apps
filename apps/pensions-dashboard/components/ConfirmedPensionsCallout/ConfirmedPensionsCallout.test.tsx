import { render } from '@testing-library/react';

import {
  EmployerStatus,
  MatchType,
  PensionType,
  UsageType,
} from '../../lib/constants';
import { ConfirmedPensionsCallout } from './ConfirmedPensionsCallout';

import '@testing-library/jest-dom/extend-expect';

const mockData = {
  schemeName: 'Test Scheme',
  employmentMembershipPeriods: [
    {
      employerName: 'Employer 1',
      employerStatus: EmployerStatus.C,
      membershipStartDate: '2010-01-01',
      membershipEndDate: '2020-01-01',
    },
    {
      employerName: 'Employer 2',
      employerStatus: EmployerStatus.H,
      membershipStartDate: '2000-01-01',
      membershipEndDate: '2010-01-01',
    },
  ],
  pensionAdministrator: {
    name: 'Admin Name',
    preferred: false,
    contactMethods: [
      {
        preferred: true,
        contactMethodDetails: {
          number: '+123 1111111111',
          usage: ['M' as UsageType],
        },
      },
    ],
  },
  externalAssetId: '12345',
  pensionType: PensionType.SP,
  benefitIllustrations: [],
  retirementDate: '2025-01-01',
  dateOfBirth: '1965-01-01',
  externalPensionPolicyId: 'policy123',
  matchType: MatchType.DEFN,
  contactReference: 'contact123',
  contributionsFromMultipleEmployers: false,
};

jest.mock('@maps-react/hooks/useTranslation', () => ({
  __esModule: true,
  default: () => ({
    t: (key: string) => key,
    locale: 'en',
  }),
}));

describe('ConfirmedPensionsCallout', () => {
  it('renders correctly', () => {
    const { container } = render(<ConfirmedPensionsCallout data={mockData} />);

    expect(container).toMatchSnapshot();
  });

  it.each([
    ['state pension type', mockData],
    ['non-state pension type', { ...mockData, pensionType: PensionType.DC }],
  ])('renders correctly with %s', (_, data) => {
    const { getByText } = render(<ConfirmedPensionsCallout data={data} />);

    expect(getByText('Test Scheme')).toBeInTheDocument();
    expect(getByText('Admin Name')).toBeInTheDocument();
    expect(getByText('common.retirement-age')).toBeInTheDocument();
    expect(getByText('common.estimated-income')).toBeInTheDocument();
    expect(getByText('common.details-link')).toBeInTheDocument();
  });

  it('renders employment membership periods', () => {
    const { getByText } = render(<ConfirmedPensionsCallout data={mockData} />);

    expect(getByText('Employer 1')).toBeInTheDocument();
    expect(getByText('Employer 2')).toBeInTheDocument();
  });

  it.each([
    ['state pension', mockData, '67'],
    ['non-state pension', { ...mockData, pensionType: PensionType.DC }, '60'],
  ])('renders the correct retirement age for %s', (_, data, expectedAge) => {
    const { getByText } = render(<ConfirmedPensionsCallout data={data} />);

    expect(getByText('common.retirement-age')).toBeInTheDocument();
    expect(getByText(expectedAge)).toBeInTheDocument();
  });

  it.each([
    ['estimated income', 'common.estimated-income', 'common.a-month'],
    ['details link', 'common.details-link'],
    ['scheme name', 'Test Scheme'],
    ['pension type', 'Admin Name'],
    ['retirement date', 'common.retirement-age', '67'],
    ['employer name', 'Employer 1', 'Employer 2'],
  ])('renders the correct %s', (_, ...expectedTexts) => {
    const { getByText } = render(<ConfirmedPensionsCallout data={mockData} />);

    expectedTexts.forEach((text) => {
      expect(getByText(text)).toBeInTheDocument();
    });
  });
});
