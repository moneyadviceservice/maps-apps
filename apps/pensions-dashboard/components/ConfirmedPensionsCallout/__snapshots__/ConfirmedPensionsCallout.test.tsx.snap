// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmedPensionsCallout renders correctly 1`] = `
<div>
  <div
    class="t-information-callout border border-slate-400 rounded-bl-3xl px-6 py-6"
    data-testid="information-callout"
  >
    <h4
      class="text-xl md:text-2xl text-blue-800 font-bold flex gap-2 mt-0 mb-4 items-top"
    >
      Test Scheme
    </h4>
    <p
      class="mb-4"
      data-testid="paragraph"
    >
      Admin Name
    </p>
    <p
      class="mb-0 text-gray-400"
      data-testid="paragraph"
    >
      common.employer
      :
    </p>
    <p
      class="pb-2 mb-2 border-b-1 border-slate-400"
      data-testid="paragraph"
    >
      Employer 1
    </p>
    <p
      class="pb-2 mb-2 border-b-1 border-slate-400"
      data-testid="paragraph"
    >
      Employer 2
    </p>
    <p
      class="mb-0 text-gray-400"
      data-testid="paragraph"
    >
      common.retirement-age
    </p>
    <p
      class="pb-2 mb-2 border-b-1 border-slate-400"
      data-testid="paragraph"
    >
      67
    </p>
    <p
      class="mb-0 text-gray-400"
      data-testid="paragraph"
    >
      common.estimated-income
    </p>
    <p
      class="mb-4 pb-2 border-b-1 border-slate-400"
      data-testid="paragraph"
    >
      <strong
        class="mr-2 text-xl"
      >
        common.amount-unavailable
      </strong>
      common.a-month
    </p>
    <a
      class="border-0 text-white bg-pink-600 hover:bg-pink-800 hover:shadow-none active:bg-pink-900 active:outline-yellow-200 active:text-white t-button px-3 py-2 font-semibold rounded outline-none shadow-bottom-gray inline-flex items-center justify-between gap-2 cursor-pointer disabled:cursor-not-allowed disabled:border-slate-400 disabled:border disabled:bg-transparent disabled:text-gray-500 disabled:shadow-none focus:bg-yellow-200 focus:outline-purple-700 focus:outline-[3px] focus:outline-offset-0 focus:text-gray-800 focus:shadow-none active:shadow-none active:outline-[3px] active:outline-offset-0"
      data-testid="details-link"
      href="/en/pension-details/12345"
    >
      common.details-link
    </a>
  </div>
</div>
`;
